src/python/modules/ensembl/compara/citest/tabletest.py

```diff
@@ -0,0 +1,245 @@
+            ## if not then failure and stop here the group by check
+            result={}
+            result["succes"]=0
+            if len(lst_result_ref) == len(lst_result_tar):
+                result["succes"]=1
+            str_info="grp:"+str_group_by+"|ref:"+str(len(lst_result_ref))+"|tar:"+str(len(lst_result_tar))
+            result["info"]=str_info
+            result["table"]=str_table_name
+            result["type"]="nb_row_all_grp"
+            self.test_results.append(result)
+            if result["succes"]==0:#if the number of group is not the same then there is no reason to continue
+                return
+
+            #load the number of autoiruied vairaiton (if group not describe in json then the variaiton expected is 0)
+            dic_autorised_variation=self._get_autorised_variation_by_group(lst_result_ref, lst_variations)
+
+            ### for aech group check that the numbers do nat variate more than autorised.
+            i=0
+            while i < len(lst_result_ref):
+                int_variation=dic_autorised_variation[lst_result_ref[i][1]]#get the aurrised variation for the group
+                result={}
+                result["succes"]=0
+                if abs(lst_result_ref[i][0]-lst_result_tar[i][0])<=int_variation: #if the number of each group is different
+                    result["succes"]=1
+                str_info="grp:"+str_group_by+"|grp_ref:"+str(lst_result_ref[i][1])+"|nb_ref:"+str(lst_result_ref[i][0])
+                str_info=str_info+"|grp_tar:"+str(lst_result_tar[i][1])+"|nb_tar:"+str(lst_result_tar[i][0])+"|var:"+str(abs(lst_result_ref[i][0]-lst_result_tar[i][0]))
+
+                result["info"]=str_info
+                result["table"]=str_table_name
+                result["type"]="nb_row_by_grp"
+                self.test_results.append(result)
+                i=i+1
+
+
+    def _get_number_row_by_group(self,obj_connection_engine,str_table_name,str_group_by):
+        lst_result=[]
+        str_request="select count(*),"+str_group_by+" from "+str_table_name+" group by "+str_group_by+";"
+        obj_request=text(str_request)
+        obj_connection=obj_connection_engine.connect()
+        obj_result=obj_connection.execute(obj_request)
+        for tpl_res in obj_result:
+            lst_result.append(tpl_res)
+        obj_connection.close()
+        return sorted(lst_result,key=lambda tuple: tuple[1])# return a list sorted on the group_by value
+
+
+    def _test_table_by_column(self, str_table_name, lst_columns):
+
+        # transfornm the json based column in sinmple stroing list.
+        lst_col =[]
+        for column in lst_columns:
+            lst_col.append(column["name"])
+
+        dic_column_ref=self._get_column_values(self.ref_connection_engine, str_table_name, lst_col)
+        dic_column_tar=self._get_column_values(self.target_connection_engine, str_table_name, lst_col)
+
+        for str_column, lst_fields_ref in dic_column_ref.items():
```

Comment:
A comment describing what's happening in this `for` loop would be useful

Reply:
I think I missed this one, it is now corrected in my code